streaming/storage/deque.go

```diff
@@ -0,0 +1,383 @@
+	dequeFirstElementKey = []byte("first")
+	dequeValueKeyPrefix  = []byte("value")
+	defaultValues        = map[string]int{
+		string(dequeFirstElementKey): 0,
+		string(dequeLastElementKey):  1,
+	}
+)
+
+type Deque struct {
+	tx           StateTransaction
+	initialized  bool
+	lastElement  int
+	firstElement int
+}
+
+type DequeIterator struct {
+	it Iterator
+}
+
+func NewDeque(tx StateTransaction) *Deque {
+	return &Deque{
+		tx:           tx,
+		initialized:  false,
+		lastElement:  1,
+		firstElement: 0,
+	}
+}
+
+func NewDequeIterator(it Iterator) *DequeIterator {
+	return &DequeIterator{
+		it: it,
+	}
+}
+
+//Adds new element to the front of the queue
+func (ll *Deque) PushFront(value proto.Message) error {
+	if !ll.initialized {
+		err := ll.initialize()
+		if err != nil {
+			return errors.Wrap(err, "failed to initialize in queue.PushFront")
+		}
+	}
+
+	data, err := proto.Marshal(value)
+	if err != nil {
+		return errors.Wrap(err, "couldn't serialize given value")
+	}
+
+	return ll.appendBytes(data, ll.firstElement, ll.decFirst)
+}
+
+//Adds new element to the end of the queue
+func (ll *Deque) PushBack(value proto.Message) error {
+	if !ll.initialized {
+		err := ll.initialize()
+		if err != nil {
+			return errors.Wrap(err, "failed to initialize in queue.PushBack()")
+		}
+	}
+
+	data, err := proto.Marshal(value)
+	if err != nil {
+		return errors.Wrap(err, "couldn't serialize given value")
+	}
+
+	return ll.appendBytes(data, ll.lastElement, ll.incLast)
+}
+
+func (ll *Deque) appendBytes(data []byte, index int, modifier func() error) error {
+	byteKey := getIndexKey(index)
+
+	err := ll.tx.Set(byteKey, data)
+	if err != nil {
+		return errors.Wrap(err, "couldn't add the element to the queue")
+	}
+
+	err = modifier()
+	if err != nil {
+		return errors.Wrap(err, "failed to actualize queue attributes")
+	}
+
+	return nil
+}
+
+//Removes the first element from the queue and returns it
+//Returns ErrEmptyQueue if the list is empty
+func (ll *Deque) PopFront(value proto.Message) error {
+	//there is no need to call initialize here, since PeekFront calls initialize
+	err := ll.PeekFront(value)
+	if err != nil {
+		return err
+	}
+
+	return ll.pop(value, ll.firstElement+1, ll.incFirst)
+}
+
+//Removes the last element from the queue and returns it
+//Returns ErrEmptyQueue if the list is empty
+func (ll *Deque) PopBack(value proto.Message) error {
+	//there is no need to call initialize here, since PeekFront calls initialize
+	err := ll.PeekBack(value)
+	if err != nil {
+		return err
+	}
+
+	return ll.pop(value, ll.lastElement-1, ll.decLast)
+}
+
+func (ll *Deque) pop(value proto.Message, index int, modifier func() error) error {
+	err := ll.peekIndex(value, index)
+	if err != nil {
+		return err
+	}
+
+	key := getIndexKey(index)
+
+	err = ll.tx.Delete(key)
+	if err != nil {
+		return errors.New("failed to delete element from queue in pop")
+	}
+
+	err = modifier()
+	return errors.Wrap(err, "failed to actualize queue attributes in pop")
```

Comment:
Difference to appendBytes function:
```
err = modifier()
if err != nil {
  return errors.Wrap(err, "failed to actualize queue attributes")
}

return nil
```

(Personally, don't really like wrapping possible nils :P . But let's just stick to 1 convention.)

Reply:
Still you haven't corrected this specific place :P